macros/contexts/contextUnits.pl

```diff
@@ -0,0 +1,2056 @@
+Use C<addUnitsNotAliases()> in place of C<addUnits()> to add just the
+named units without adding any aliases for them.
+
+=head2 Custom units
+
+You can define your own units in terms of the fundamental units.
+E.g., to define the unit C<acres>, you could use
+
+    Context("Units")->withUnitsFor("length")->addUnits(acres => {factor => 4046.86, m => 2});
+
+which indicates that 1 acre is equal to 4046.86 square meters.
+
+You can even make up your own fundamental units.  For example, to
+define C<apples> and C<oranges> as units, you could do
+
+    Context("Units")->addUnits(
+      apples => { apples => 1, aliases => ["apple"] },
+      oranges => { oranges => 1, aliases => ["orange"] }
+    );
+
+    BEGIN_PGML
+    If you have 5 apples and give your friend 2 of them,
+    what do you have left? [___________]{"3 apples"}
+    END_PGML
+
+and the student can answer C<3 oranges> but will be marked incorrect
+(with a message about the units being incorrect).  Note that C<apples>
+and C<apple> are synonymous in this context, and that C<1 apple> is
+accepted, but is displayed as C<1 apples>, as no attempt is made to
+handle plurals.
+
+On the other hand, you could also do
+
+    Context("Units")->addUnits(
+      apples => { fruit => 1, aliases => ["apple"] },
+      oranges => { fruit => 1, aliases => ["orange"] }
+    );
+
+    Compute("3 apples") == Compute("3 oranges"); # returns 1
+
+will consider apple and oranges as the same unit (both are the
+fundamental unit of C<fruit>).
+
+Finally,
+
+    Context("Units")->addUnits(
+      apples => { fruit => 1, aliases => ["apple"] },
+      oranges => { fruit => 1, aliases => ["orange"], factor => 2 }
+    );
+
+    Compute("1 apple") == Compute("2 oranges"); # returns 1
+
+will make an apple equivalent to two oranges by making both C<apples>
+and C<oranges> be examples of the fundamental unit C<fruit>.
+
+You can remove individual units from the context using the
+C<removeUnits()> method of the context.  For example
+
+    Context("Units")->withUnitsFor("length")->removeUnits("ft", "inch", "mile", "furlong");
+
+removes the English units and their aliases, leaving only the metric
+units.  To remove a unit without removing its aliases, use C<removeUnitsNotAliases()>
+instead.
+
+Note that the units are stored in the context as constants, so to list
+all the units, together with other constants, use
+
+    Context()->constants->names;
+
+The constants that are units have the C<isUnit> property set.  So
+
+    grep {Context()->constants->get($_)->{isUnit}} (Context()->constants->names);
+
+will get the list of units.
+
+
+=head2 Adding units to other contexts
+
+The C<Units> and C<LimitedUnits> contexts are based on the C<Numeric>
+and C<LimitedNumeric> contexts.  You can add units to other contexts
+using the C<context::Units::extends()> function.  For example,
+
+    Context(context::Units::extending("Fraction")->withUnitsFor("length"));
```

Comment:
To use this example, one also must include the 'contextFraction.pl' macro.  I think we may need to make a comment about this.

Reply:
fixed